src/core/net/dns_headers.hpp

```diff
@@ -881,6 +1065,70 @@ class PtrRecord : public ResourceRecord
+     * This method also verifies that the PTR record is well-formed (e.g., the record data length `GetLength()` matches
+     * the PTR encoded name).
+     *
+     * @param[in]     aMessage          The message to read from.  `aMessage.GetOffset()` MUST point to the start of
+     *                                  DNS header.
+     * @param[inout]  aOffset           On input, the offset in @p aMessage to start of PTR name field.
+     *                                  On exit when successfully read, @p aOffset  is updated to point to the byte
+     *                                  after the entire PTR record (skipping over the record).
+     * @param[out]    aNameBuffer       A pointer to a char array to output the read name as a null-terminated C string
+     *                                  (MUST NOT be nullptr).
+     * @param[in]     aNameBufferSize   The size of @p aNameBuffer.
+     *
+     * @retval OT_ERROR_NONE            The PTR name was read successfully. @p aOffset and @p aNameBuffer are updated.
+     * @retval OT_ERROR_PARSE           The PTR record in @p aMessage could not be parsed (invalid format).
+     * @retval OT_ERROR_NO_BUFS         Name could not fit in @p aNameBufferSize chars.
+     *
+     */
+    otError ReadPtrName(const Message &aMessage, uint16_t &aOffset, char *aNameBuffer, uint16_t aNameBufferSize) const
+    {
+        return ResourceRecord::ReadName(aMessage, aOffset, /* aStartOffset */ aOffset - sizeof(PtrRecord), aNameBuffer,
+                                        aNameBufferSize,
+                                        /* aSkipRecord */ true);
+    }
+
+    /**
+     * This method parses and reads the PTR name from a message.
+     *
+     * This method also verifies that the PTR record is well-formed (e.g., the record data length `GetLength()` matches
+     * the PTR encoded name).
+     *
+     * Unlike the previous method which reads the entire PTR name into a single char buffer, this method reads the
+     * first label separately and into a different buffer @p aLabelBuffer and the rest of the name into @p aNameBuffer.
+     * The @p aNameBuffer can be set to `nullptr` if the caller is only interested in the first label. This method is
+     * intended for "Service Instance Name" where first label (`<Instance>` portion) can be a user-friendly string and
+     * can contain dot character.
+     *
+     * @param[in]     aMessage          The message to read from. `aMessage.GetOffset()` MUST point to the start of
+     *                                  DNS header.
+     * @param[inout]  aOffset           On input, the offset in @p aMessage to the start of PTR name field.
+     *                                  On exit, when successfully read, @p aOffset  is updated to point to the byte
+     *                                  after the entire PTR record (skipping over the record).
+     * @param[out]    aLabelBuffer      A pointer to a char array to output the first label as a null-terminated C
```

Comment:
Is the first label supposed to be the instance name?

However, I think the instance name can contain `.` character. Should we correctly handle such case?

Reply:
FYI, in #6103, I introduced a method to split the full names into \<Instance\>.\<Service\>.\<Protocol\>.\<Domain\>, which would allow us to better handle a full service/instance name as a null-terminated string.

Reply:
> Is the first label supposed to be the instance name?
> 
> However, I think the instance name can contain . character. Should we correctly handle such case?
> 

That's exactly the reason we have this method which reads the first label separately ... 😃 
I mention this in the comment/description just few lines above this (at line 1101).

> FYI, in #6103, I introduced a method to split the full names into <Instance>.<Service>.<Protocol>.<Domain>, which would allow us to better handle a full service/instance name as a null-terminated string.

My suggestion is to read and treat the "instance name" label seperately (that's what we do in the SPR client code and what I will do in DNS-SD client code).

If we have a full name as a C string with `'.'` separating the labesl, then we need to be able to tell if a dot `.` char is part of the instance name or it is a separator. We would need use some escaping mechanism. I think reading and keeping instance name label seperately as its own C string may be simpler.

Reply:
Sorry I got it wrong. 

Let's discuss the it further in DNS-SD Server PR if necessary.

Reply:
No worries at all. 😃